internal/daemon/api_logs.go

```diff
@@ -0,0 +1,326 @@
+		response.ServeHTTP(w, req)
+		return
+	}
+	follow := followStr == "true"
+
+	numLogs := defaultNumLogs
+	nStr := query.Get("n")
+	if nStr != "" {
+		n, err := strconv.Atoi(nStr)
+		if err != nil {
+			response := statusBadRequest("n must be a valid integer")
+			response.ServeHTTP(w, req)
+			return
+		}
+		numLogs = n
+	}
+
+	// If "services" parameter not specified, fetch logs for all services.
+	if len(services) == 0 {
+		infos, err := r.svcMgr.Services(nil)
+		if err != nil {
+			response := statusInternalError("cannot fetch services: %v", err)
+			response.ServeHTTP(w, req)
+			return
+		}
+		services = make([]string, len(infos))
+		for i, info := range infos {
+			services[i] = info.Name
+		}
+	}
+
+	// Get log iterators by service name (and close them when we're done).
+	itsByName, err := r.svcMgr.ServiceLogs(services, numLogs)
+	if err != nil {
+		response := statusInternalError("cannot fetch log iterators: %v", err)
+		response.ServeHTTP(w, req)
+		return
+	}
+	defer func() {
+		for _, it := range itsByName {
+			_ = it.Close()
+		}
+	}()
+
+	// Output format is not exactly text/plain, but it's not pure JSON either.
+	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+	if numLogs != 0 {
+		if len(itsByName) == 1 {
+			// Handle single-service case efficiently
+			for name, it := range itsByName {
+				outputLogsSingle(w, name, it)
+			}
+		} else if numLogs < 0 {
+			// n<0, output all logs from all services in time order
+			outputLogsAll(w, itsByName)
+		} else {
+			outputLogsMulti(w, itsByName, numLogs)
+		}
+	}
+	if follow {
+		flushWriter(w)
+		followLogs(w, itsByName, req.Context().Done())
+	}
+}
+
+// Efficiently output a single iterator's logs.
+func outputLogsSingle(w io.Writer, service string, it servicelog.Iterator) {
+	now := time.Now().UTC() // stop if we iterate past current time
+	for it.Next() && it.Timestamp().Before(now) {
+		err := writeLog(w, service, it.Timestamp(), it.StreamID(), it.Length(), it)
+		if err != nil {
+			fmt.Fprintf(w, "\ncannot write log from %q: %v", service, err)
+			return
+		}
+	}
+}
+
+// Output last numLogs logs from multiple iterators (less efficient as it
+// writes them all to a slice first, and sorts by timestamp).
+func outputLogsMulti(w io.Writer, itsByName map[string]servicelog.Iterator, numLogs int) {
+	type entry struct {
+		timestamp time.Time
+		service   string
+		stream    servicelog.StreamID
+		message   string
+	}
+
+	// Write all entries to slice.
+	var entries []entry
+	var buf bytes.Buffer
+	now := time.Now().UTC() // stop if we iterate past current time
+	for name, it := range itsByName {
+		for it.Next() && it.Timestamp().Before(now) {
+			buf.Reset()
+			_, err := io.Copy(&buf, it)
+			if err != nil {
+				fmt.Fprintf(w, "\ncannot write log from %q: %v", name, err)
+				return
+			}
+			entries = append(entries, entry{
+				timestamp: it.Timestamp(),
+				service:   name,
+				stream:    it.StreamID(),
+				message:   buf.String(),
```

Comment:
we should return to this shortly with reverse iterators on the write buffer.

Imagine 10 (let m=10) services, 1mb logs each, using `pebble logs -n 10000` (let n=10000). It's going to allocate another ~10mb just to then sort through `n*m=100000` logs. Worst case that's 10mb and `O(m*n^2)=10b` iterations, best case 100kb and `O(m*n*log(m*n))=500k` iterations.

With reverse iterators we could do this in no extra memory, `O(2n)=200k` iterations.

Reply:
Yeah, agreed it's quite expensive at present. Per offline discussion, I'm totally fine if you do this later as a separate PR.